processing: 3.5.3 -> 3.5.4 | fix / refactor

9999years opened the pull request:
The old expression downloaded `http://download.processing.org/reference.zip` with a fixed hash, and would break when new versions were released.

This expression remedies that by using a versioned tarball.

Further, it avoids compiling by applying our patch (to avoid the spurious "Not fond of this Java VM" warning message) directly to the compiled `.class` file.

###### Motivation for this change

The old `processing` expression was broken by a change to `http://download.processing.org/reference.zip`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

9999years committed:
processing: 3.5.3 -> 3.5.4 | fix / refactor

The old expression downloaded
`http://download.processing.org/reference.zip` with a fixed hash, and
would break when new versions were released.

This expression remedies that by using a versioned tarball.

Further, it avoids compiling by applying our patch (to avoid the
spurious "Not fond of this Java VM" warning message) directly to the
compiled `.class` file.

lourkeur commented:
Builds and launches well on NixOS x86_64 provided an oraclejdk8. Some of my sketches have issues with libraries, but that's still better than status quo. Nicely done with the bytecode patch. LGTM

bbjubjub2494 reviewed (approved):
Also works on `i686-linux` afaict

lourkeur commented on pkgs/applications/graphics/processing/default.nix:
```diff
@@ -1,58 +1,133 @@
-{ stdenv, fetchFromGitHub, fetchurl, xmlstarlet, makeWrapper, ant, jdk, rsync, javaPackages, libXxf86vm, gsettings-desktop-schemas }:
-
-stdenv.mkDerivation rec {
+{ stdenv, lib, jdk, patchelf, makeWrapper, xlibs, zip, unzip, rsync
```
[1/2]
```suggestion
{ stdenv, lib, jdk, patchelf, makeWrapper, xlibs, zip, unzip, rsync, fetchzip
```
[`fetchTarball`](https://nixos.org/nix/manual/#ssec-builtins) is meant to pull Nix expressions at evaluation time. If Nixpkgs `fetchzip` is used instead it should help with the OfBorg error.

lourkeur commented on pkgs/applications/graphics/processing/default.nix:
```diff
@@ -1,58 +1,133 @@
-    repo = "processing";
-    rev = "processing-0269-${version}";
-    sha256 = "0ajniy3a0i0rx7is46r85yh3ah4zm4ra1gbllmihw9pmnfjgfajn";
+  src = fetchTarball {
```
```suggestion
  src = fetchzip {
```
[2/2]

9999years committed:
processing: s/fetchTarball/fetchzip, s/xlibs/xorg

Co-authored-by: Louis Bettens <lourkeur@users.noreply.github.com>

Use fetchzip instead of fetchTarball

Co-authored-by: Louis Bettens <lourkeur@users.noreply.github.com>

Use xorg instead of xlibs

Use xorg instead of xlibs

Use xorg instead of xlibs

lourkeur commented:
[s/xlibs/xorg/g](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/aliases.nix#L625)

9999years commented:
Want me to squash those nit fixes up?

lourkeur commented:
Yes I think it'd be nice.

9999years committed:
processing: make bytecode patch overridable

Abstracting out values from the bytecode patch / making it
toggleable guards against future changes breaking the derivation.

9999years committed:
processing: fix function arg order

- Arguments to `lib.map` were reversed.
- `mkDerivation` complained about stringifying a set; use variables
  prefixed with `popup` instead.

9999years commented:
OK, got those bugs fixed.
